### PR TITLE
refactor(client): Refactor scalar arguments validation

### PR DESCRIPTION
**Pull request by SevInf**
After few dives into `query.ts` I believe we could handle scalar arguments validation a little bit better. 

What we are doing right now: for every passed in value we try to guess it's type, based on typeof, schema type, matching certain regexps, etc. From that we build a stringified name of the type and compare with expected name, built from the schema. In a long list of conditionals then we try to see if the types are compatible (you can assign ints to float types, but not vice-versa) as well as correct some incorrect "guesses" from previous step (if a string looks like a date we'd report that it's a date, but then allow to pass dates to string types if we guessed incorrectly).

This has a couple of issues:
1. Since we compare types as strings, lists have to be processed separately from scalars. If we say that `Foo` is compatible with `Bar`, we need to also not forget that `List<Foo>` is compatible with `List<Bar>`, it won't happen automatically. It causes issues like #11674 
2. Need to correct incorrect guesses, like in String-Date example from above.

What I am proposing instead in this PR:

1. Given DMMF schema for types, we build a series of classes for each supported scalar type, as well as the list. 
2. Each class is responsible for checking, if provided value is compatible. We don't try to guess the type from the value. If user passes us a string that looks like a `DateTime`, we won't say it is a `DateTime` unless it's used in a position where `DateTime` is an acceptable input.
3. List type reuses the classes of scalar types, so if we implement changes in particular scalar type logic, lists of those scalars will handle this case as well.

I believe this a clearer and less susceptible to errors approach.  It does not seem to break any existing tests. Plus, comment above `hasCorrectScalarType` said to refactor it :)

This PR is for gathering initial feedback on this from people who know the codebase better than I do: do you think it's viable approach? Does it actually improve things? Am I missing some edge cases?

If we actually decide to ship it, it will need much more tests, including tests for `queryArgsTypecheck` module itself, hence the draft. I also did not remove `getGraphQLType` "guessing" function, since we seem to use it for error messages and it still might be useful.

### CODE REVIEW

File: packages/client/src/runtime/queryArgsTypecheck.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    return new ListType(type)
+  }
+
+  return type
+}
+
+function typeByName(name: string): Type {
+  const TypeConstructor = dmmfTypesMapping[name]
+  if (!TypeConstructor) {
+    throw new TypeError(`Unknown scalar type: ${name}`)
```
Comment (SevInf) [inline packages/client/src/runtime/queryArgsTypecheck.ts:174]: Question: do we have a specific error type for the cases of "This should never have happened in real life, but it did and here we are"?

**Comment by SevInf** [comment]
JSON protocol supersedes this
